diff-tests/core/model/src/properties.ts

```diff
@@ -0,0 +1,256 @@
+  };
+
+  /**
+   *
+   * @param chain chain of block
+   * @param height height of block
+   * @returns Returns the height of the least successing block on the counterparty
+   * chain if it exists, else undefined.
+   */
+  getLeastSucc = (chain: Chain, height: number): number | undefined => {
+    const it = this.leastSucc[chain].keys();
+    let bestH = 100000000000000; // Infinity
+    let bestAnswer = -1;
+    let result = it.next();
+    while (!result.done) {
+      const h = result.value;
+      if (h < bestH && height <= h) {
+        bestH = h;
+        bestAnswer = this.leastSucc[chain].get(h) as number;
+      }
+      result = it.next();
+    }
+    if (bestAnswer === -1) {
+      // No least successor exists.
+      return undefined;
+    }
+    return bestAnswer;
+  };
+}
+
+class BlockHistory {
+  partialOrder = new PartialOrder();
+  blocks: Record<Chain, Map<number, CommittedBlock>> = {
+    provider: new Map(),
+    consumer: new Map(),
+  };
+  /**
+   * Mark state as permanently committed to the blockchain.
+   * @param chain
+   * @param invariantSnapshot
+   */
+  commitBlock = (chain: Chain, invariantSnapshot: InvariantSnapshot) => {
+    const h = invariantSnapshot.h[chain];
+    const b: CommittedBlock = {
+      chain,
+      invariantSnapshot,
+    };
+    this.blocks[chain].set(h, b);
+  };
+}
+
+function sum(arr: number[]): number {
+  return arr.reduce((sum: number, x: number) => sum + x, 0);
+}
+
+/**
+ * Is the total fund value in the system constant?
+ * It only makes sense to check this if slashing with non-zero
+ * slash factors never occurs, because slashing with non-zero
+ * slash factors burns funds.
+ *
+ * @param hist A history of blocks.
+ * @returns Is the property satisfied?
+ */
+function stakingWithoutSlashing(hist: BlockHistory): boolean {
+  const blocks = Array.from(hist.blocks[P].entries())
+    .sort((a, b) => a[0] - b[0])
+    .map((e) => e[1])
+    .map((b) => b.invariantSnapshot);
+
+  function value(e: InvariantSnapshot) {
+    let x = e.delegatorTokens;
+    x += sum(e.tokens);
+    x += sum(e.undelegationQ.map((e) => e.balance));
+    return x;
+  }
+
+  const v = value(blocks[0]);
+  for (let i = 1; i < blocks.length; i++) {
+    if (value(blocks[i]) !== v) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ * Checks the bond-based consumer voting power property as defined
+ * in https://github.com/cosmos/ibc/blob/main/spec/app/ics-028-cross-chain-validation/system_model_and_properties.md#system-properties
+ * but modified to account for finite executions and always zero slash factors.
+ *
+ * @param hist A history of blocks.
+ * @returns Is the property satisfied?
+ */
+function bondBasedConsumerVotingPower(hist: BlockHistory): boolean {
+  const partialOrder = hist.partialOrder;
+  const blocks = hist.blocks;
+  function powerProvider(block: CommittedBlock) {
+    return _.range(NUM_VALIDATORS).map(
+      (i) =>
+        block.invariantSnapshot.tokens[i] +
```

Comment:
These are all the tokens of validator `i` regardless of whether it is bonded, unbonding, or unbonded. For the property, only `bonded` and `unbonding` validators are relevant (plus the unbonding undelegations that are captured in the next lines).

Reply:
Useful catch!

https://github.com/cosmos/interchain-security/blob/07fa9a82c8ed9a93baa42e74a46c7df00349a42b/diff-tests/core/model/src/properties.ts#L189-L208